pridefetch: init at 1.0.0

Minion3665 opened the pull request:
###### Description of changes

- Add myself to the maintainer-list.nix
- Create the pridefetch package at [version 1.0.0](https://github.com/SpyHoodle/pridefetch/releases/tag/v1.0.0)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Minion3665 committed:
maintainers: add minion3665

Mindavi reviewed (changes requested):
I don't think we accept maintainer additions from people that aren't planning on being a nixpkgs maintainer.

Minion3665 committed:
pridefetch: init at 1.0.0

pridefetch is a cli tool to display system information (similarly
to neofetch or pfetch) in your terminal along with a pride flag.

Minion3665 commented:
> I don't think we accept maintainer additions from people that aren't planning on being a nixpkgs maintainer.

Alright; I'll add the package that I have made for pridefetch to nixpkgs as well as planned, although I will do so in this PR rather than separate PRs

Mindavi commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,45 @@
+    maintainers = [
+      maintainers.minion3665
+    ];  # TODO: Replace this with a reference to the maintainer list after adding myself to it
+    platforms = systems.supported.hydra;
```
We have something like platforms.all or platforms.unix

Mindavi commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,45 @@
+    rev = "v" + version;
+    inherit sha256;
+  };
+in stdenv.mkDerivation {
```
Could `buildPython*` be used here maybe?

Minion3665 replied:
I don't think so, as pridefetch doesn't have stuff needed for a python package such as `setup.py`. It's an idea for the future, but it wouldn't be able to be used on the current version

Mindavi commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,45 @@
+    ]))
+    zip
+  ];
+  unpackPhase = "true";
```
dontUnpack = true

Mindavi commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,45 @@
+  installPhase = ''
+    mkdir -p $out/bin
+    cd ${src}/src
+    zip -r $out/pridefetch.zip *
```
I might be tired but this whole installPhase looks kind of strange to me...

Minion3665 replied:
You're right; I've changed it from all being in installPhase to being partly in installPhase and partly in buildPhase. installPhase now only moves the completed executable to the bin directory, whereas buildPhase zips the files and adds the header so the program can be executed by python

Mindavi commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,45 @@
+      The flag which is printed is configurable, as well as the width of the output.
+    '';
+    homepage = https://github.com/SpyHoodle/pridefetch;
+    license = "bsd";
```
We have licenses that can be referred to in the lib, please use those instead of strings.

ncfavier commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,46 @@
+{ stdenv, fetchFromGitHub, lib, python39, zip }: let
```
`lib` usually comes first. Also, I think it's better not to depend on a specific minor version of Python. `python3` should be fine.

ncfavier commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,46 @@
+    inherit sha256;
+  };
+in stdenv.mkDerivation {
+  inherit name version;
```
You want to pass a `pname` here, not `name`.

ncfavier commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,46 @@
+    ]))
+    zip
+  ];
+  dontUnpack = true;
```
This doesn't seem necessary: remove this line, add `src` to the derivation, and just do `cd src`.

ncfavier commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,46 @@
+    (python39.withPackages (pythonPackages: with pythonPackages; [
+      distro
+    ]))
+    zip
```
`zip` is used at build time and should go in `nativeBuildInputs`.

ncfavier commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,46 @@
+      Pridefetch prints your system statistics (similarly to neofetch, screenfetch or pfetch) along with a pride flag.
+      The flag which is printed is configurable, as well as the width of the output.
+    '';
+    homepage = https://github.com/SpyHoodle/pridefetch;
```
Unquoted URL syntax is [deprecated](https://github.com/NixOS/rfcs/pull/45), please use double quotes.

ncfavier commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,46 @@
+  ];
+  dontUnpack = true;
+  buildPhase = ''
+    mkdir $out
```
Typically build products aren't moved to `$out` in the build phase.

Here's my proposed division of the work between the two phases (I'd also be fine with putting everything in `installPhase`, given that I can't decide whether "zipping a directory and slapping a shebang on top" qualifies as "build" or "install"):

- in the build phase, create the `pridefetch` executable in the current directory (and make it executable).

- in the install phase, `mkdir -p $out/bin` and move `pridefetch` there.

And whatever you do, don't forget to add [phase hooks](https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/missing-phase-hooks.md).

ncfavier reviewed (commented):
A few remarks. https://github.com/jtojnar/nixpkgs-hammering can help find some of them.

Minion3665 commented:
> A few remarks. [jtojnar/nixpkgs-hammering](https://github.com/jtojnar/nixpkgs-hammering) can help find some of them.

Thank you very much for your help and suggesting use of nixpkgs-hammering! I decided to keep the build step and install step separate, as I think that the build step here creates the executable and the install step installs it

ncfavier commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,52 @@
+  ];
+  buildPhase = ''
+    runHook preBuild
+    builddir=$(pwd)
+    pushd src
+    zip -r $builddir/pridefetch.zip ./*
```
```suggestion
    pushd src
    zip -r ../pridefetch.zip .
```

Relative paths are fine here but in general, prefer `$PWD` to `$(pwd)`.

ncfavier reviewed (commented):
Final nit!

Mindavi commented on pkgs/tools/misc/pridefetch/default.nix:
```diff
@@ -0,0 +1,51 @@
+    pushd src
+    zip -r ../pridefetch.zip ./*
+    popd
+    echo '#!/usr/bin/env python' | cat - pridefetch.zip > pridefetch
```
Is this fixed by patchShebangs? Otherwise the package may not work at all if python is not installed. Maybe try with nix-shell --pure to see if it works well in that environment?

Mindavi replied:
Also, the dance with zipping still looks a bit weird to me. Why is it needed again?

ncfavier replied:
> Is this fixed by patchShebangs?

Yes

Minion3665 replied:
> Also, the dance with zipping still looks a bit weird to me. Why is it needed again?

we have multiple python files in our project, and we can't just add a shebang to the top of our entry point file and be done with it as otherwise either the structure would change or we'd have to move all of the files into the bin directory. Zipping allows us to have a single file that contains all of the code and can be executed by python

Mindavi commented:
Already looks quite good, some small things to look at and then I think it'll be ok for now.

Mindavi commented:
This looks fine. Maybe a simple installCheck test could be added in a followup, if you'd like.
